keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,521 @@
+**Note:** *Not required until targeted at a release.*
+
+Define graduation milestones.
+
+These may be defined in terms of API maturity, or as something else. The KEP
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
+or by redefining what graduation means.
+
+In general we try to use the same stages (alpha, beta, GA), regardless of how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- N examples of real-world usage
+- N installs
+- More rigorous forms of testing—e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least two releases between beta and
+GA/stable, because there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include 
+[conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+### Upgrade / Downgrade Strategy
+
+#### Handling of the "`kubernetes`" Service
+
+kube-apiserver attempts to force the "`kubernetes`" Service definition
+to match its expectations on startup, but does not attempt to
+reconcile it after that point.
+
+Thus, in a cluster with both old single-stack-only and new dual-stack
+apiservers, whichever apiserver started last will determine whether
+the Service is currently single or dual stack. In a simple upgrade or
+downgrade, this would result in the Service definition changing when
+the first apiserver was upgraded/downgraded, and then not changing
+after that. In a more complicated upgrade/downgrade scenario, there
+might be some flapping; eg, the first master upgrades and changes the
+Service to dual-stack, but then the second master reboots before
+upgrading, causing the old apiserver to restart and revert the Service
+definition back to its single-stack form, before then being upgraded
+and changing it back to dual-stack again.
+
+However, at all times, the Service should fully usable; it should
+never claim to be dual-stack unless there actually are dual-stack
```

Comment:
a dual-stack service will always be fully usable if the primary ip family matches the single-stack family, right?

Reply:
I'm not talking about generic promises we make about dual-stack Services, I'm just talking about the specific temporary states of the apiserver service during this particular upgrade.

And actually, there's a case I wasn't thinking about: because the Service is only changed at apiserver startup, but the Endpoints get resynced while running, you can end up in a case where the Service is dual-stack, but there is no EndpointSlice for the secondary IP family. In that case, trying to connect to the Service's secondary IP would fail with "connection refused". Although, I guess as long as clients behave sanely, they'd cope with that anyway; this would only be a problem if some client tried to connect _only_ to the secondary-family ClusterIP for some reason, rather than trying to connect _only_ to the primary-family ClusterIP, or trying both in sequence/parallel.